### PR TITLE
Adds an example of how to use 'Arrow'

### PR DESCRIPTION
**Pull request by fagossa**
This commit adds an example of data validation using `Arrow` showing the use of:
- Applicatives (Validated)
- Monads (Either)

See https://github.com/arrow-kt/arrow

### CODE REVIEW

File: src/main/kotlin/fr/xebia/xke/5_FP.kt [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+     * - Validated.applicative<List<String>>()
+            .map2(validated1, validated2, { v: Tuple<Int, Int> -> v.a + v.b })
+            .ev()
+     * - NonEmptyList.of("a string", "another string")
+     */
+
+    // TODO find a way to aggregate in an applicative way the 'readName' and 'readAge' functions
+    fun validateUser(params: Map<String, String>): Validated<NonEmptyList<String>, User>  {
+        return Validated.applicative<NonEmptyList<String>>()
+            .map2(
```
Comment (sergiords) [inline src/main/kotlin/fr/xebia/xke/5_FP.kt]: - Don't understand how we can extend this mechanism to 3 or more checks.
- Here we have a List of errors, ok, but each validation `readName` and `readAge` should only return one error not a list of errors. Typing should provide an hint about expected behaviour.

File: src/main/kotlin/fr/xebia/xke/5_FP.kt [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    // TODO read the field 'age' and validate that is non blank, valid int, and non negative
+    fun readAge(params: Map<String, String>): Either<NonEmptyList<String>, Int> =
+        getValue("age", params)
+            .flatMap { nonBlank ("age", it) }
+            .flatMap { parseInt ("age", it) }
+            .flatMap { nonNegative (it) }
+
+    // TODO get a value from the map or return an error
+    fun getValue(name: String, data: Map<String, String>): Either<NonEmptyList<String>, String> =
+        Option.fromNullable(data[name]).toEither { NonEmptyList.of("$name is not present") }
```
Comment (sergiords) [inline src/main/kotlin/fr/xebia/xke/5_FP.kt]: I prefer `data[name]?.let { Either.right(it) } ?: Either.left("$name is not present")`

File: build.gradle [hunk 3]
```diff
@@ -17,6 +17,13 @@ dependencies {
     compile "org.jetbrains.kotlin:kotlin-stdlib-jre8"
     compile "org.jetbrains.kotlinx:kotlinx-coroutines-core:0.18"
     compile "com.fasterxml.jackson.module:jackson-module-kotlin:2.9.0"
+    compile 'io.arrow-kt:arrow-core:0.6.1'
+    compile 'io.arrow-kt:arrow-typeclasses:0.6.1'
+    compile 'io.arrow-kt:arrow-data:0.6.1'
+    compile 'io.arrow-kt:arrow-instances:0.6.1'
+    compile 'io.arrow-kt:arrow-syntax:0.6.1'
+    kapt    'io.arrow-kt:arrow-annotations-processor:0.6.1'
```
Comment (sergiords) [inline build.gradle]: Features are still poorly documented, can we illustrate this FP pattern without external dependency?

File: src/main/kotlin/fr/xebia/xke/5_FP.kt [hunk 4]
```diff
@@ -0,0 +1,71 @@
+package fr.xebia.xke
+
+import arrow.core.*
+import arrow.syntax.option.toEither
+
+import arrow.core.Either
+import arrow.core.Option
+import arrow.core.flatMap
+import arrow.data.NonEmptyList
```
[the remark this answers is not in the document]
Comment (fagossa) [reply src/main/kotlin/fr/xebia/xke/5_FP.kt]: Actually it is needed; the `accumulation` with applicatives only works with `NonEmptyList`

File: src/main/kotlin/fr/xebia/xke/5_FP.kt [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+        try { Either.right(data.toInt()) }
+        catch(e: NumberFormatException) {
+            Either.left(NonEmptyList.of("$name must be an integer"))
+        }
+
+    fun nonBlank(name: String, data: String): Either<NonEmptyList<String>, String> =
+        Either.cond(data.isNotEmpty(), { data }, { NonEmptyList.of("$name is blank") })
+
+    fun nonNegative(data: Int): Either<NonEmptyList<String>, Int> =
+        Either.cond(data > 0, { data }, { NonEmptyList.of("negative value") })
```
[the remark this answers is not in the document]
Comment (fagossa) [reply src/main/kotlin/fr/xebia/xke/5_FP.kt]: Why not ...

File: src/main/kotlin/fr/xebia/xke/5_FP.kt (hunk 2 above)
Comment (fagossa) [reply src/main/kotlin/fr/xebia/xke/5_FP.kt]: Ok, why not

File: build.gradle (hunk 3 above)
Comment (fagossa) [reply build.gradle]: The amount of code needed to be able to do replicate the same behaviour could be considered a completely project/exercise.

Even if the doc in lacking `Arrow` is considered the most complete FP library for kotlin.

File: src/main/kotlin/fr/xebia/xke/5_FP.kt [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+        Option.fromNullable(data[name]).toEither { NonEmptyList.of("$name is not present") }
+
+    fun parseInt(name: String, data: String): Either<NonEmptyList<String>, Int> =
+        try { Either.right(data.toInt()) }
+        catch(e: NumberFormatException) {
+            Either.left(NonEmptyList.of("$name must be an integer"))
+        }
+
+    fun nonBlank(name: String, data: String): Either<NonEmptyList<String>, String> =
+        Either.cond(data.isNotEmpty(), { data }, { NonEmptyList.of("$name is blank") })
```
[the remark this answers is not in the document]
Comment (fagossa) [reply src/main/kotlin/fr/xebia/xke/5_FP.kt]: OK, why not

File: src/main/kotlin/fr/xebia/xke/5_FP.kt [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+     * - Validated.applicative<List<String>>()
+            .map2(validated1, validated2, { v: Tuple<Int, Int> -> v.a + v.b })
+            .ev()
+     * - NonEmptyList.of("a string", "another string")
+     */
+
+    // TODO find a way to aggregate in an applicative way the 'readName' and 'readAge' functions
+    fun validateUser(params: Map<String, String>): Validated<NonEmptyList<String>, User>  {
+        return Validated.applicative<NonEmptyList<String>>()
+            .tupled(
```
Comment (fagossa) [inline src/main/kotlin/fr/xebia/xke/5_FP.kt]: I replaced `map2` by `tupled` to avoid confusion

File: src/main/kotlin/fr/xebia/xke/5_FP.kt (hunk 1 above)
Comment (fagossa) [reply src/main/kotlin/fr/xebia/xke/5_FP.kt]: I'm not inventing nothing new here!
this is the way it is typically done in several FP languages

File: src/main/kotlin/fr/xebia/xke/5_FP.kt (hunk 7 above)
Comment (sergiords) [reply src/main/kotlin/fr/xebia/xke/5_FP.kt]: This is the method I was looking for in order to check 3 or more fields

Comment (sergiords) [reply src/main/kotlin/fr/xebia/xke/5_FP.kt]: This is the method I was looking for in order to check 3 or more fields

File: src/main/kotlin/fr/xebia/xke/5_FP.kt (hunk 1 above)
Comment (sergiords) [reply src/main/kotlin/fr/xebia/xke/5_FP.kt]: No I haven't, I was asking how we could apply check to 3 or more fields :)
Use of `applicative.tupled` in the last comment answers this question

File: build.gradle (hunk 3 above)
Comment (sergiords) [reply build.gradle]: I get your point, however I do not think attendees (beginner to intermediate) will have time to dig in arrow implementation details, to find appropriate methods (like I unsuccessfully did).
We could, as an alternative, implement a simple validation mechanism gathering all field errors.

Comment (fagossa) [reply build.gradle]: I agree, thats why I added a comment containing useful functions. That might not be enough for beginners though (:sob:). I propose the following alternatives:

* Create another markdown (or even in the readme) giving more details on the kind of problem vs function to use. Some sections could include (applicatives, monads, disjunctions, etc)
* Improve the comment section in `UserValidation` and maybe a proper `Kdoc` showing how to use the functions (map2, tupled, left, right, etc)

File: src/main/kotlin/fr/xebia/xke/5_FP.kt [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+/**
+ * This class validates user data in two ways:
+ * - Applicative for all fields (validated all fields)
+ * - Monadic within each field (stops on the first error per field)
+ *
+ *
+ * The following functions might help you
+ * - Option.fromNullable("something")
+ * - Option(maybeRight).toEither { "Left value" }
+ * - Try { data.toInt() }.toEither().mapLeft { NonEmptyList.of("$name must be an integer") }
```
Comment (fagossa) [inline src/main/kotlin/fr/xebia/xke/5_FP.kt]: Even if the `Try` monad exist, I'll leave it as a possible alternative and not as part of the solution

File: build.gradle (hunk 3 above)
Comment (fagossa) [reply build.gradle]: Actually I just added to the `kdoc` of the file

File: src/main/kotlin/fr/xebia/xke/5_FP.kt [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+ * - Either.cond(true, { 42 }, { "Error" })
+ * - Validated.fromEither(...)
+ * - Validated.applicative(NonEmptyList.semigroup<String>())
+ *     .map2(validated1, validated2, { v: Tuple<Int, Int> -> v.a + v.b })
+ *     .fix()
+ * - Validated.applicative(NonEmptyList.semigroup<String>())
+ *    .tupled(validated1, validated2, ..., validatedN)
+ *    .fix()
+ *    .map{ v: Tuple<Int, Int> -> v.a + v.b }
```
Comment (sergiords) [inline src/main/kotlin/fr/xebia/xke/5_FP.kt]: ` .map{ v: TupleN<Int, ..., Int> -> v.a + ... + v.b }`

**Pull request merged**